src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/CompareSymbolsCorrectlyAnalyzer.cs

```diff
@@ -38,15 +44,61 @@ public override void Initialize(AnalysisContext context)
             {
                 var compilation = context.Compilation;
                 var symbolType = compilation.GetTypeByMetadataName(s_symbolTypeFullName);
+
                 if (symbolType is null)
                 {
                     return;
                 }
 
                 context.RegisterOperationAction(context => HandleBinaryOperator(in context, symbolType), OperationKind.BinaryOperator);
+
+                var comparerTypes = s_comparerTypeFullName
+                    .Select(comparerTypeFullName => compilation.GetTypeByMetadataName(comparerTypeFullName))
+                    .WhereNotNull()
+                    .ToImmutableArray();
+
+                if (comparerTypes.IsDefaultOrEmpty)
+                {
+                    return;
+                }
+
+                context.RegisterOperationAction(context => HandleOperation(in context, comparerTypes), OperationKind.ObjectCreation);
             });
         }
 
+        private void HandleOperation(in OperationAnalysisContext context, ImmutableArray<INamedTypeSymbol> comparerTypes)
+        {
+            if (context.Operation is IObjectCreationOperation objectCreation)
+            {
+                var typeBeingCreated = (INamedTypeSymbol)objectCreation.Type;
+
+                // Does the type have constructors that take comparers? 
+                var constructorsWithComparers = typeBeingCreated.Constructors.WhereAsArray(constructor => hasSymbolComparer(constructor, comparerTypes));
+
+                if (constructorsWithComparers.Length == 0)
+                {
+                    return;
+                }
+
+                // Is the constructor being used one of the ones that takes a comparer? 
+                var correctConstructorBeingUsed = constructorsWithComparers.Contains(objectCreation.Constructor);
+
+                if (!correctConstructorBeingUsed)
+                {
+                    context.ReportDiagnostic(objectCreation.Syntax.GetLocation().CreateDiagnostic(Rule));
+                }
+            }
+
+            // Local functions
+
+            static bool hasSymbolComparer(IMethodSymbol methodSymbol, ImmutableArray<INamedTypeSymbol> comparerTypes)
```

Comment:
PascalCase for local functions.

```suggestion
            static bool HasSymbolComparer(IMethodSymbol methodSymbol, ImmutableArray<INamedTypeSymbol> comparerTypes)
```